src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -13288,6 +13079,114 @@ private void LeaveQuery()
             _syntaxFactoryContext.QueryDepth--;
         }
 
+        private delegate (SyntaxToken openToken, PostSkipAction action) SkipBadTokens<TNode>(
+            LanguageParser parser, SyntaxToken openToken, SeparatedSyntaxListBuilder<TNode> builder, SyntaxKind expectedKind, SyntaxKind closeTokenKind) where TNode : GreenNode;
+
+        private SeparatedSyntaxList<TNode> ParseCommaSeparatedSyntaxList<TNode>(
+            ref SyntaxToken openToken,
+            SyntaxKind closeTokenKind,
+            Func<LanguageParser, bool> isPossibleElement,
+            Func<LanguageParser, TNode> parseElement,
+            SkipBadTokens<TNode> skipBadTokens,
+            bool allowTrailingSeparator,
+            SyntaxKind expectedKind = SyntaxKind.IdentifierToken,
+            ErrorCode? trailingSeparatorError = null) where TNode : GreenNode
+        {
+            return ParseSeparatedSyntaxList<TNode>(
+                ref openToken,
+                SyntaxKind.CommaToken,
+                closeTokenKind,
+                isPossibleElement,
+                parseElement,
+                skipBadTokens,
+                allowTrailingSeparator,
+                expectedKind,
+                trailingSeparatorError);
+        }
+
+        private SeparatedSyntaxList<TNode> ParseSeparatedSyntaxList<TNode>(
+            ref SyntaxToken openToken,
+            SyntaxKind separatorTokenKind,
+            SyntaxKind closeTokenKind,
+            Func<LanguageParser, bool> isPossibleElement,
+            Func<LanguageParser, TNode> parseElement,
+            SkipBadTokens<TNode> skipBadTokens,
+            bool allowTrailingSeparator,
+            SyntaxKind expectedKind = SyntaxKind.IdentifierToken,
+            ErrorCode? trailingSeparatorError = null) where TNode : GreenNode
+        {
```

Comment:
behold the glory.

Reply:
It is a big improvement to have `goto tryAgain` get consolidated into here. I still find this method a bit difficult as a reader. I wouldn't demand it, but do wonder if we can revise it so that it's easier to grasp what it's doing and why.

For example, the switch expression arm parsing code, while it sounds like it might have its own issues (@alrz mentioned?), I found *vastly* easier to understand in structure.

Reply:
This is to the point, BTW, that I need to set aside time later this afternoon to review this function specifically before I can sign off, everything else looks great.

Reply:
Definitely agree.  I'm keeping the same for now so that it's easy to see it's the same.  but anyone who wants to clean this up only has this single place to do so :)